### PR TITLE
KVM Ingestion - Import Instance

### PR DESCRIPTION
**Pull request by kishankavala**
### Description

This PR adds new functionality to import KVM instance from an external host or from disk image in local or shared storage

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [X] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@kishankavala 

can you fix the packaging issue, getting the following exception 

```
21:05:08 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project cloud-server: Compilation failure
21:05:08 [ERROR] /jenkins/workspace/acs-centos7-pkg-builder/dist/rpmbuild/BUILD/cloudstack-4.19.0.0-SNAPSHOT/server/src/main/java/org/apache/cloudstack/vm/UnmanagedVMsManagerImpl.java:[1605,45] error: method getUnmanagedNicNetworkMap in class UnmanagedVMsManagerImpl cannot be applied to given types;
21:05:08 [ERROR]   required: String,List<Nic>,Map<String,Long>,Map<String,IpAddresses>,DataCenter,String,Account
21:05:08 [ERROR]   found: List<Nic>,Map<String,Long>,Map<String,IpAddresses>,DataCenter,String,Account
21:05:08 [ERROR]   reason: actual and formal argument lists differ in length
21:05:08 [ERROR] 
```

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by kishankavala** [comment]
@blueorangutan package

**Comment by kishankavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan test

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Review by kiranchavala — APPROVED** [review]
LGTM, Tested the feature manually and its working fine

**Comment by kiranchavala** [comment]
@nvazquez can you review and approve the PR

**Comment by kiranchavala** [comment]
@harikrishna-patnala  can you review and approve the PR

**Review by harikrishna-patnala — APPROVED** [review]
I've made some minor indentation changes and reverted the EVENT_VM_CREATE on deploy VM change which is not looking good with the events. cc @nvazquez @kishankavala 

After all these this PR code LGTM

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan test

**Review by nvazquez — COMMENTED** [review]
I still think we can improve this PR by reducing the amount of repeated code, proposed changes on:

https://github.com/apache/cloudstack/pull/7976/files#r1338547526
https://github.com/apache/cloudstack/pull/7976/files#r1338549482
https://github.com/apache/cloudstack/pull/7976/files#r1338558081
https://github.com/apache/cloudstack/pull/7976/files#r1338560072

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan ui

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 1]
```diff
@@ -5179,4 +5205,100 @@ public void setInterfaceDefQueueSettings(Map<String, String> details, Integer cp
...
+                s_logger.debug("VM " + dm.getName() + ": powerstate = " + ps + "; vm state=" + state.toString());
+                final String vmName = dm.getName();
+
+                if (state == PowerState.PowerOff) {
+                    UnmanagedInstanceTO unmanagedInstanceTO = new UnmanagedInstanceTO();
+                    unmanagedInstanceTO.setName(vmName);
+                    unmanagedInstanceTO.setPowerState(UnmanagedInstanceTO.PowerState.PowerOff);
+                    unmanagedInstanceTO.setCpuCores(getCpuShares(dm));
+                    //unmanagedInstanceTO.setMemory(new Long(dm.getMaxMemory()).intValue());
+                    unmanagedInstanceTO.setMemory(131072);
```
[the remark this answers is not in the document]
Comment (kishankavala) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java]: Removed fixed value

File: ui/src/views/compute/wizard/MultiNetworkSelection.vue [hunk 2]
```diff
@@ -197,12 +197,6 @@ export default {
       this.validNetworks = {}
       for (const item of this.items) {
         this.validNetworks[item.id] = this.networks
-        if (this.filterUnimplementedNetworks) {
```
[the remark this answers is not in the document]
Comment (kishankavala) [reply ui/src/views/compute/wizard/MultiNetworkSelection.vue]: Fixed

File: ui/src/views/tools/ImportUnmanagedInstance.vue [hunk 3]
```diff
@@ -671,7 +684,21 @@ export default {
...
+          hypervisor: this.hypervisor,
+          url: this.hostname,
+          username: this.username,
+          password: this.password,
+          hostid: this.host.id,
+          storageid: this.pool.id,
+          diskpath: this.diskpath
+        }
+        api = 'importUnmanagedInstance'
+        if (this.importsource === 'extrenal' || this.importsource === 'local' || this.importsource === 'shared') {
```
[the remark this answers is not in the document]
Comment (kishankavala) [reply ui/src/views/tools/ImportUnmanagedInstance.vue]: fixed

File: server/src/main/java/org/apache/cloudstack/vm/UnmanagedVMsManagerImpl.java [hunk 4]
```diff
@@ -494,7 +501,8 @@ private StoragePool getStoragePool(final UnmanagedInstanceTO.Disk disk, final Da
     private Pair<UnmanagedInstanceTO.Disk, List<UnmanagedInstanceTO.Disk>> getRootAndDataDisks(List<UnmanagedInstanceTO.Disk> disks, final Map<String, Long> dataDiskOfferingMap) {
         UnmanagedInstanceTO.Disk rootDisk = null;
         List<UnmanagedInstanceTO.Disk> dataDisks = new ArrayList<>();
-        if (disks.size() == 1) {
+        //ToDo: fix cdrom
```
[the remark this answers is not in the document]
Comment (kishankavala) [reply server/src/main/java/org/apache/cloudstack/vm/UnmanagedVMsManagerImpl.java]: Fixed

File: ui/src/views/tools/ImportUnmanagedInstance.vue [hunk 5]
```diff
@@ -499,6 +508,10 @@ export default {
       }
       return meta
     },
+    isLocal () {
```
[the remark this answers is not in the document]
Comment (kishankavala) [reply ui/src/views/tools/ImportUnmanagedInstance.vue]: Fixed

File: core/src/main/java/com/cloud/agent/api/GetRemoteVmsCommand.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+package com.cloud.agent.api;
+
+@LogLevel(LogLevel.Log4jLevel.Trace)
+public class GetRemoteVmsCommand extends Command {
```
[the remark this answers is not in the document]
Comment (kishankavala) [reply core/src/main/java/com/cloud/agent/api/GetRemoteVmsCommand.java:23]: GetRemoteVmsCommand  has no fields in common with GetUnmanagedInstancesCommand

File: core/src/main/java/com/cloud/agent/api/GetRemoteVmsAnswer.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+package com.cloud.agent.api;
+
+import org.apache.cloudstack.vm.UnmanagedInstanceTO;
+
+import java.util.HashMap;
+import java.util.List;
+
+@LogLevel(LogLevel.Log4jLevel.Trace)
+public class GetRemoteVmsAnswer extends Answer {
```
[the remark this answers is not in the document]
Comment (kishankavala) [reply core/src/main/java/com/cloud/agent/api/GetRemoteVmsAnswer.java:26]: unmanagedInstances is common in both Answers. But its better to have 1:1 mapping with command and answer

File: api/src/main/java/org/apache/cloudstack/api/command/admin/vm/ImportVmCmd.java [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+
+@APICommand(name = "importVm",
+        description = "Import virtual machine from a unmanaged host",
+        responseObject = UserVmResponse.class,
+        responseView = ResponseObject.ResponseView.Full,
+        requestHasSensitiveInfo = false,
+        responseHasSensitiveInfo = true,
+        authorized = {RoleType.Admin},
+        since = "4.19.0")
+public class ImportVmCmd extends BaseAsyncCmd {
```
[the remark this answers is not in the document]
Comment (kishankavala) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/vm/ImportVmCmd.java:66]: ImportUnmanagedInstanceCmd  already has 15 parameters. Not all of them are supported by ImportVmCmd. There are another 8 more parameters in ImportVmCmd. Some consolidation in the implemenatation can be done in next phase

**Comment by shwstppr** [comment]
@kishankavala @nvazquez @kiranchavala is this ready or still under work?

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test matrix

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by kishankavala** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by andrijapanicsb** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@kishankavala build is failing here. Can you please check and run packaging, tests later?

**Comment by kishankavala** [comment]
@blueorangutan package

**Comment by kishankavala** [comment]
@blueorangutan test

**Comment by kishankavala** [comment]
@blueorangutan package

**Comment by kishankavala** [comment]
@blueorangutan package

**Comment by kishankavala** [comment]
@blueorangutan package

**Comment by kishankavala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by kishankavala** [comment]
@blueorangutan package

**Review by andrijapanicsb — APPROVED** [review]
LGTM, after detailed testing.

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan test

**Pull request merged**